### PR TITLE
Added pseudo code

### PR DESCRIPTION
**Pull request by systemzax**
Updated pseudo code : libraries/chain/hotstuff/hs_pseudo

### CODE REVIEW

**Review by greg7mdp — APPROVED** [review]

File: libraries/chain/hotstuff/hs_pseudo [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+	incremental_block_mtree		finality_mtree;
+	uint32_t					policy_generation;
+}
+
+struct block_handle {
+	block_header_state_ptr 						bhs;
+	finalizer_policy_ptr 						fp;	
+	std::optional<pending_quorum_certificate>	pending_qc; 
+	std::optional<valid_quorum_certificate>		valid_qc; 
+	uint64_t id() const {return bhs->id;}
```
Comment (heifner) [inline libraries/chain/hotstuff/hs_pseudo:90]: `uint64_t` -> `block_id_type` which is `sha256`

File: libraries/chain/hotstuff/hs_pseudo [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+	sha256          proposal_digest; //proposal digest
+	bls_public_key  finalizer_key;
+	bls_signature   sig;
+	bool 			weak; //indicate if vote is weak, strong otherwise
+};
+
+
+//added as a block_header extension before signing
+struct hotstuff_header_extension {
+	uint32_t					last_qc_block_height;
```
Comment (heifner) [inline libraries/chain/hotstuff/hs_pseudo:151]: Not sure why this is only the height. Doesn't it need to be block id or block digest?

File: libraries/chain/hotstuff/hs_pseudo [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+}
+
+//when a node receives a vote on a proposal
+void on_vote_received(hs_vote_message v){
+	//check for duplicate or invalid vote, return in either case
+	//abstracted [...]
+	block_handle& bc = fork_db.get_block_by_id(v.block_id);
+	am_i_leader = [...] //abstracted, must return true if I am the leader, false otherwise
+	if(!am_i_leader) return;
+	//only leader need to take action on votes
```
Comment (heifner) [inline libraries/chain/hotstuff/hs_pseudo:318]: Need to relay votes even if not a leader.

File: libraries/chain/hotstuff/hs_pseudo (hunk 2 above)
Comment (systemzax) [reply libraries/chain/hotstuff/hs_pseudo:151]: I think for the header extension we only store the last_qc_block_height and is_last_qc_strong.

Assuming the block is valid, we should always be able to locate an ancestor block at last_qc_block_height on that branch and get its id. 

See : 
https://github.com/AntelopeIO/leap/issues/1911

File: libraries/chain/hotstuff/hs_pseudo (hunk 3 above)
Comment (systemzax) [reply libraries/chain/hotstuff/hs_pseudo:318]: I added a comment to mention this. Also added a similar comment on_proposal_received for consistency.

**Comment by ericpassmore** [comment]
Note:start
group: IF
category: DESIGN
summary: Proposed data-structures and terminology for new faster finality. 
Note:end

**Pull request merged**